### DISCUSSION TITLE
What's the best way to format a program that uses Condorcet?

### DISCUSSION THREAD

**Discussion by LiamM32** [julien-boudry/Condorcet#142] [Q&A] [answered]
I'm sorry to bother you with some amateur questions on using PHP. But I've never worked with PHP before working with Condorcet. What I know comes from reading w3schools.com & php.net, just so I can work with Condorcet.

I am trying to make a program to process an election with a specialized variant of the Schulze method. This method is too specialized to ever be worth including in Condorcet itself. How should I format a program that requires Condorcet as an API?

First I started by writing a new php file in `src/Algo/Schulze/`, which would include this method. I also put 2 .json files there which it uses for data.

But then I figured that's not a good place to put it, especially for when I share this. So I put a new subfolder in `Examples/`, and tried writing a new php file there. But the problem is that php files in this folder don't know how to access CondorcetPHP. I include the line `use CondorcetPHP\Condorcet\{Condorcet, Candidate, Election, Vote};`, but PHP doesn't recognize this. This is true even if I add the lines `require_once __DIR__.'/../../__CondorcetAutoload.php';` and/or `include __DIR__.'/../../__CondorcetAutoload.php';`.

But perhaps the ideal way to format this project is as new project that includes Condorcet as a library. Anyone who downloads it can download Condorcet separately. But I don't know how to get a PHP program to use an external library. I know there is [this page](https://www.condorcet.io/3.AsPhpLibrary/1.Installation), but I don't know how to use composer. I also don't know how to install Condorcet as a PHP library on my computer. I am on Manjaro Linux.

So how should I do this? What would the beginning of the initial PHP file look like?

**Comment by julien-boudry** [comment]
You're welcome :-) It's very rare to interact with real users, and really useful to realize some imperfections or bugs. It's a niche program. And when it works well and easily, people (most of the time) don't say anything, they're ghosts.

This is a good point, I will write a little skeleton ASAP about creating a method module without fork Condorcet.

Condorcet is designed for (optionally) working without composer and dependencies for security concerns (except for the Command line version), but I highly recommend using it with Composer because is nicer and you can include it with another project more easily. It also activates an arbitrary precision floating point Math library, useful for some methods (better than PHP float precision).

About Composer, it's not a big deal: https://getcomposer.org/

**Reply by LiamM32** [reply]
Thank you. You have clearly put much more effort into documentation than what I have found to be typical for small open-source projects. Actually, it's better than the macro documentation for Libreoffice, which is nearly non-existent. I have gotten more assistance from you than I have gotten with my previous attempts to contribute to open-source software projects.

Though the documentation is still a few steps away from being suited to inexperienced amateur programmers, or people not used to PHP. This isn't something I expect from all open-source libraries. But perhaps it would be useful to have a clear example of a php file for a program that uses the Condorcet API.

**Reply by LiamM32** [reply]
A problem I've had is writing a new method that extends Schulze_Core. I haven't figured out how to make it work without putting it in the same folder. I have been getting errors that I haven't managed to solve.

**Comment by LiamM32** [comment]
It looks like you've started making a [reference example](https://github.com/CondorcetVote/Condorcet_Modules_Skeletons). Thank you.

**Comment by LiamM32** [comment]
I don't know where else to post this.

@julien-boudry I have tried downloading your [Condorcet_Modules_Skeletons](https://github.com/CondorcetVote/Condorcet_Modules_Skeletons). The program works properly when I do the steps in README.md. However, I don't know how to rename things without breaking it. You can see [this repository](https://github.com/LiamM32/Condorcet_Modules_Skeletons_failure), which is the same, except some namespaces and the method class are renamed. Somehow the program no longer works with these changes.

```
[liam@Liam-Manjaro-Desktop Condorcet_Modules_Skeletons]$ php src/main.php
PHP Fatal error:  Uncaught CondorcetPHP\Condorcet\Throwable\AlgorithmException: The voting algorithm is not available: no class found for 'FancyVoting\Method\Fancy_Method' in /home/liam/src/Condorcet_Modules_Skeletons/vendor/julien-boudry/condorcet/src/Condorcet.php:189
Stack trace:
#0 /home/liam/src/Condorcet_Modules_Skeletons/vendor/julien-boudry/condorcet/src/Condorcet.php(170): CondorcetPHP\Condorcet\Condorcet::testMethod()
#1 /home/liam/src/Condorcet_Modules_Skeletons/src/main.php(11): CondorcetPHP\Condorcet\Condorcet::addMethod()
#2 {main}
  thrown in /home/liam/src/Condorcet_Modules_Skeletons/vendor/julien-boudry/condorcet/src/Condorcet.php on line 189
```

**Reply by LiamM32** [reply]
I just managed to fix this by renaming the class from `Fancy_Method` to `FancyMethod`, matching the file name. I'll leave that repository up for a few hours in case you have anything to comment. Then I'll delete it, as I seem to have figured out the problem.

**Comment by julien-boudry** [comment, marked as answer]
@LiamM32 Just improve the skeleton with PHPUnit test, command line version, and other cleanups.

https://github.com/CondorcetVote/Condorcet_Modules_Skeletons
